src/main/java/com/google/sps/data/PersistentUserRepository.java

```diff
@@ -0,0 +1,130 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+
+/**
+ * This class implements the match repository using the datastore 
+ */
+public class PersistentUserRepository implements UserRepository {
+    
+  private Collection<User> allUsers = new ArrayList<>();
+  
+  private DatastoreService datastore;
+  
+  private int maxProfiles = 10;
+  
+  public PersistentUserRepository() {
+      datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  // function to add profile to database
+  public void addUserToDatabase(User user) {
+    String name = user.toString();
+    Collection<String> specialties = user.getSpecialties();
+
+    Entity userEntity = new Entity("User");
+    userEntity.setProperty("name", name);
+    if(specialties.size() > 0){
+        userEntity.setProperty("specialties", specialties);
+    }
+    datastore.put(userEntity);
+  }
+
+  // function to fetch the user from the database
+  public Collection<User> fetchUserProfiles() {
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> resultsList = results.asList(FetchOptions.Builder.withLimit(maxProfiles));
+
+    Collection<User> userProfiles = new ArrayList<>();
+    for (Entity entity : resultsList) {
+        String name = (String) entity.getProperty("name");
+        Collection<String> specialties = (Collection) entity.getProperty("specialties");
+
+        User userObject = new User(name);
+        if(specialties != null) {
+            for (String specialty : specialties) {
+                userObject.addSpecialty(specialty);
+            }
+        }
+        userProfiles.add(userObject);
+    }
+    return userProfiles;
+  }
+
+  public void removeUserProfile(User user){
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> resultsList = results.asList(FetchOptions.Builder.withLimit(maxProfiles));
+
+    String name = user.toString();
+
+    for (Entity entity : resultsList) {
+        String currentName = (String) entity.getProperty("name");
+        if(name.contains(currentName)) {
+            Key current = entity.getKey();
+            datastore.delete(current);
+        }
+    }
+  }
+  // function that adds user to the database if it does not exist already
+  public void addUser(User user) {
+    allUsers = fetchUserProfiles();
+    /*
+     * if the user exists in the database then nothing happens,
+     * if it doesnt exist then it gets added
+    */
+    if(!allUsers.contains(user)) {
+        allUsers.add(user);
+        addUserToDatabase(user);
+    }
+
+  }
+
+  public void removeUser(User user) throws Exception {
+    if(allUsers.contains(user)) {
+        allUsers.remove(user);
+        removeUserProfile(user);
+    } else {
+        throw new Exception("Can't remove user that does not exist");
+    }
+  }
+
+  public Collection<User> getAllUsers() {
+    Collection<User> results = fetchUserProfiles();
```

Comment:
fetchUserProfiles() only fetches 10 profiles. does that match the expected behavior for getAllUsers()?

Reply:
What do you mean?

Reply:
Yes it now matches and should not have a limit on the amount of profiles being received

Reply:
Done